### PR TITLE
testCacheManager: use cppunit exception tests

### PR DESCRIPTION
**Pull request by kinkie**
Do not hand-roll tests for exception-throwing code

### CODE REVIEW

**Review by rousskov — APPROVED** [review]

**Review by yadij — APPROVED** [review]

**Comment by kinkie** [comment]
> If possible, please apply similar changes to TestRandomUuid::testInvalidIds() and check main() in test_http_range.cc

Noted and added to backlog

**Comment by kinkie** [comment]
> > If possible, please apply similar changes to TestRandomUuid::testInvalidIds() and check main() in test_http_range.cc
> 
> Noted and added to backlog

testRandomUuid done in PR #1814 , test_http_range is not applicable (it's simply reporting on unexpected exceptions, rather than enforcing that an exception is thrown)

**Review by rousskov — COMMENTED** [review]
> test_http_range is not applicable (it's simply reporting on unexpected exceptions ...)

I disagree that test_http_range is not applicable. Its main() is not just reporting (that it is _dying_). It is exiting with an error code on any thrown exception:

```C++
    {
        ...
    } catch (const std::exception &e) {
        printf("Error: dying from an unhandled exception: %s\n", e.what());
        return EXIT_FAILURE;
    } catch (...) {
        printf("Error: dying from an unhandled exception.\n");
        return EXIT_FAILURE;
    }
```

Would not EXIT_FAILURE fail "make check"?

> ... rather than enforcing that an exception is thrown

AFAICT, test_http_range is enforcing (via EXIT_FAILURE) that an exception is _not_ thrown. Just like CPPUNIT_ASSERT_NO_THROW() code added in this PR. Am I missing something?

**Comment by kinkie** [comment]
> > test_http_range is not applicable (it's simply reporting on unexpected exceptions ...)
> 
> I disagree that test_http_range is not applicable. Its main() is not just reporting (that it is _dying_). It is exiting with an error code on any thrown exception:
> 
> ```c++
>     {
>         ...
>     } catch (const std::exception &e) {
>         printf("Error: dying from an unhandled exception: %s\n", e.what());
>         return EXIT_FAILURE;
>     } catch (...) {
>         printf("Error: dying from an unhandled exception.\n");
>         return EXIT_FAILURE;
>     }
> ```
> 
> Would not EXIT_FAILURE fail "make check"?
> 
> > ... rather than enforcing that an exception is thrown
> 
> AFAICT, test_http_range is enforcing (via EXIT_FAILURE) that an exception is _not_ thrown. Just like CPPUNIT_ASSERT_NO_THROW() code added in this PR. Am I missing something?

Of course EXIT_FAILURE makes "make check" fail. So does not catching an exception.
The difference with other test cases is that here we are not validating that some code
raises an exception we would expect it to do; here we are saying "if an exception is thrown, fail the test".
Duh, every test does that implicitly.

Now, I didn't write this code, but I can see it was added in commit a0b3b22c8938bbd358f6e3c47b8c69d4a21fba37 with this message: "Range: header testing may thor exceptions which were not caught by the test binary. Could lead to difficulty debugging exception errors."

Which matches my interpretation of the code: "if the exception code follows the standard, print the message, it might be helpful".

In other words, I think that the best thing we can do here is to just remove the exception block and reduce this test to the default case: unhandled exceptions terminate() and signal failure to the test harness

**Review by rousskov — COMMENTED** [review]
The last [response](https://github.com/squid-cache/squid/pull/1811#issuecomment-2119634811) appears to be based on testInvalidUrl() changes. They are a red herring in this discussion because test_http_range code is _not_ similar to testInvalidUrl(). test_http_range is similar to testValidUrl(). Hence, the last response does not help me understand why test_http_range case was declared inapplicable when this PR handles the same "make sure the exception is _not_ thrown (and report it if it is)" case inside PR's CacheManagerInternals::testValidUrl().

> I think that the best thing we can do here is to just remove the exception block and reduce this test to the default case: unhandled exceptions terminate() and signal failure to the test harness

AFAICT, that proposed approach goes against the inapplicable claim, CacheManagerInternals::testValidUrl() changes in this PR, and commit a0b3b22 concerns. Please do not interpret this paragraph as implying that the proposed approach is necessarily wrong.

Let's continue this discussion in new #1816 that is dedicated to test_http_range changes.

**Review by yadij — APPROVED** [review]

**Comment by squidadm** [comment]
queued for backport to v6
